pkg/operator/encryption/statemachine/transition.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	// add new resources without keys. These resources will trigger STEP 2.
+	oldEncryptedGRs := make([]schema.GroupResource, 0, len(desiredEncryptionState))
+	for _, gr := range toBeEncryptedGRs {
+		if _, ok := desiredEncryptionState[gr]; !ok {
+			desiredEncryptionState[gr] = state.GroupResourceState{}
+		} else {
+			oldEncryptedGRs = append(oldEncryptedGRs, gr)
+		}
+	}
+
+	//
+	// STEP 1: without secrets, wait for the key controller to create one
+	//
+	// the code after this point assumes at least one secret
+	if len(encryptionSecrets) == 0 {
+		klog.V(4).Infof("no encryption secrets found")
+		return desiredEncryptionState
+	}
+
+	// override fake secrets with real ones if they exist
+	for gr, grState := range desiredEncryptionState {
+		for i, rk := range grState.ReadKeys {
+			for _, k := range backedKeys {
+				if state.EqualKeyAndEqualID(&rk, &k) {
+					grState.ReadKeys[i] = k
+					break
+				}
+			}
+		}
+		if grState.HasWriteKey() {
+			for _, s := range backedKeys {
+				if state.EqualKeyAndEqualID(&grState.WriteKey, &s) {
+					grState.WriteKey = s
+					break
+				}
+			}
+		}
+		desiredEncryptionState[gr] = grState
+	}
+
+	//
+	// STEP 2: verify to have all necessary read-keys. If not, add them, deploy and wait for stability.
+	//
+	// Note: we never drop keys here. Dropping only happens in STEP 4.
+	// Note: only keysWithPotentiallyPersistedData are considered. There might be more which are not pruned yet by the pruning controller.
+	//
+	// TODO: allow removing resources (e.g. on downgrades) and transition back to identity.
+	allReadSecretsAsExpected := true
+	currentlyEncryptedGRs := oldEncryptedGRs
+	if oldEncryptionConfig == nil {
+		// if the config is not there, we assume it was deleted. Assume worst case of all toBeEncryptedGRs were encrypted.
+		currentlyEncryptedGRs = toBeEncryptedGRs
+	}
+	expectedReadSecrets := state.KeysWithPotentiallyPersistedData(currentlyEncryptedGRs, backedKeys)
+	for gr, grState := range desiredEncryptionState {
+		changed := false
+		for _, expected := range expectedReadSecrets {
+			found := false
+			for _, rk := range grState.ReadKeys {
+				if state.EqualKeyAndEqualID(&rk, &expected) {
+					found = true
+					break
+				}
+			}
+			if !found {
+				// Just adding raw key without trusting any metadata on it
+				grState.ReadKeys = append(grState.ReadKeys, expected)
+				changed = true
+				allReadSecretsAsExpected = false
+				klog.V(4).Infof("encryped resource %s misses read key %s", gr, expected.Key.Name)
+			}
+		}
+		if changed {
+			grState.ReadKeys = state.SortRecentFirst(grState.ReadKeys)
+			desiredEncryptionState[gr] = grState
+		}
+
+		// verify that all read secrets are backed by a real secret, and therefore
+		// in encryptionSecrets. Read secrets are sorted. So encryptionSecrets[0]
+		// is a good next write key because it is most recent.
+		for _, rk := range grState.ReadKeys {
+			if !rk.Backed {
```

Comment:
interesting keys that are not backed can prevent newly added `gr` from progressing.
what if a key that is missing is actually a write key?

Reply:
The key controller should create a new key. Can you verify?

Reply:
nice! - they key will be created - https://github.com/openshift/cluster-kube-apiserver-operator/pull/606/files#diff-d1dba536a50febb495e1ce860c933aaaR290

Reply:
what about a missing read key? when it will be removed from `desiredEncryptionState`

Reply:
It is only remove in the last step. It is the only step that ever removes a read key.

Reply:
```
		// potential write-key must be backed. Otherwise stop here in STEP 2 and let key controller create a new key.
		if !grState.ReadKeys[0].Backed {
			allReadSecretsAsExpected = false
		}
```